imports/wasi_snapshot_preview1/fs.go

```diff
@@ -194,15 +194,130 @@ var fdFdstatSetRights = stubFunction(
+//		          []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 117, 80, 0, 0, 0, 0, 0, 0, 160, 153, 212, 128, 110, 221, 35, 23, 160, 153, 212, 128, 110, 221, 35, 23, 160, 153, 212, 128, 110, 221, 35, 23}
+//		resultFilestat   ^-- dev                 ^-- ino                 ^                       ^-- nlink               ^-- size                   ^-- atim                              ^-- mtim                              ^-- ctim
+//		                                                                 |
+//		                                                                 +-- filetype
+//
+// The following properties of filestat are not implemented:
+//   - dev: not supported by Golang FS
+//   - ino: not supported by Golang FS
+//   - nlink: not supported by Golang FS
+//   - atime: not supported by Golang FS, we use mtim for this
+//   - ctim: not supported by Golang FS, we use mtim for this
+//
+// Note: This is similar to `fstat` in POSIX.
 // See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-fd_filestat_getfd-fd---errno-filestat
-var fdFilestatGet = stubFunction(
-	functionFdFilestatGet,
-	[]wasm.ValueType{i32, i32},
-	[]string{"fd", "result.buf"},
+// and https://linux.die.net/man/3/fstat
+var fdFilestatGet = &wasm.HostFunc{
+	ExportNames: []string{functionFdFilestatGet},
+	Name:        functionFdFilestatGet,
+	ParamTypes:  []api.ValueType{i32, i32},
+	ParamNames:  []string{"fd", "result.buf"},
+	ResultTypes: []api.ValueType{i32},
+	Code: &wasm.Code{
+		IsHostFunction: true,
+		GoFunc:         api.GoModuleFunc(fdFilestatGetFn),
+	},
+}
+
+type wasiFiletype uint8
+
+const (
+	wasiFiletypeUnknown wasiFiletype = iota
+	wasiFiletypeBlockDevice
+	wasiFiletypeCharacterDevice
+	wasiFiletypeDirectory
+	wasiFiletypeRegularFile
+	wasiFiletypeSocketDgram
+	wasiFiletypeSocketStream
+	wasiFiletypeSymbolicLink
 )
 
+func fdFilestatGetFn(ctx context.Context, mod api.Module, params []uint64) []uint64 {
+	fd := uint32(params[0])
+	buf := uint32(params[1])
+
+	sysCtx := mod.(*wasm.CallContext).Sys
+	file, ok := sysCtx.FS(ctx).OpenedFile(ctx, fd)
+	if !ok {
+		return errnoBadf
+	}
+
+	fileStat, err := file.File.Stat()
+	if err != nil {
+		return errnoIo
+	}
+
+	fileMode := fileStat.Mode()
+
+	wasiFileMode := wasiFiletypeUnknown
+	if fileMode&fs.ModeDevice != 0 {
+		wasiFileMode = wasiFiletypeBlockDevice
+	} else if fileMode&fs.ModeCharDevice != 0 {
+		wasiFileMode = wasiFiletypeCharacterDevice
+	} else if fileMode&fs.ModeDir != 0 {
+		wasiFileMode = wasiFiletypeDirectory
+	} else if fileMode&fs.ModeType == 0 {
+		wasiFileMode = wasiFiletypeRegularFile
+	} else if fileMode&fs.ModeSymlink != 0 {
+		wasiFileMode = wasiFiletypeSymbolicLink
+	}
+
+	if !mod.Memory().WriteByte(ctx, buf+16, uint8(wasiFileMode)) {
```

Comment:
if you feel up to making this more efficient, go for it! I noticed @ckaznocha hunting down some perf and we might as well make a good example forward even if the rest of this file isn't. If this is beyond your budget one of us can do in a follow-up as generally this sort of thing is the fun part. you did the hard part :)

The key idea is that when writing fixed width data, we can use
`buf, ok = memory.Read(ctx, offset, byteCount)` to get a write-through buffer of the range. You may want to zero out any padding.

Then, we can get read the mtime only once like and convert it to a slice like this, then write it down to the slice. that or add functions in u64 to write the slice instead of making one.

e.g. in current code:
```
mtim := fileStat.ModTime().UnixNano()
mtimSlice := u64.LeBytes(mtim)

// now copy it a few times to buf you made.
```



u64

Reply:
Hey, hope I'm not over stepping but since I heard my name I took a peek 😉.

Alternately:
```suggestion
	mtim := uint64(fileStat.ModTime().UnixNano())
    binary.LittleEndian.PutUin64(buf[offset+40:], mtim)
    binary.LittleEndian.PutUin64(buf[offset+48:], mtim)
    binary.LittleEndian.PutUin64(buf[offset+56:], mtim)
```

I'm honestly not sure if it would be faster to allocate the new slice and copy it 3 times vs writing directly to the write-through buffer but doing the bitwise operations 3 times. Either way write-through buffer seems like a good call!

Reply:
I agree with your approach! in any case a bench could be added here to see the impact. I used this earlier to validate the significant perf impact when we stopped using reflective signatures https://github.com/tetratelabs/wazero/blob/2173f3075878b89162884bf5ce280d59f6212c69/imports/wasi_snapshot_preview1/wasi_bench_test.go

Reply:
Sounds good, will make the changes later! Would zero'ing out the padding do anything? Isn't the memory already zero'ed when it's being allocated?

Reply:
I'm not terribly concerned about the zeroing. good point!

Reply:
done!